### PR TITLE
Add flexible bilinear upsampling aspect ratio redux

### PR DESCRIPTION
**Pull request by andrewgiessel**
This PR addresses issue #1257, which aims to add non-coupled scaling factors for bilinear 2d upsampling. I've added two new tests, and currently everything passes. This is a relatively simple change which only touches python code.

Furthermore, this is is a re-do of PR https://github.com/pytorch/pytorch/pull/1279, wherein I messed up a rebase.  There are some in-line comments there that might be worth skimming through, as well, but I tried to address most of the concerns raised there.  In particular, I maintained the base class of all the upsampling methods.

Tests pass, on GPU and CPU.

tagging @apaszke @fmassa @soumith.  Thanks in advance everyone.

ps:  I'll probably need a pointer on how to best rebase to master, eventually ;)

### CODE REVIEW

File: torch/nn/_functions/thnn/upsampling.py [hunk 1]
```diff
@@ -12,16 +13,20 @@ def __init__(self, size=None, scale_factor=None):
...
         if size is None and scale_factor is None:
             raise ValueError('either size or scale_factor should be defined')
-        if scale_factor is not None and not isinstance(scale_factor, Integral):
-            raise ValueError('scale_factor must be of integer type')
-        if size is not None and not isinstance(size, tuple):
-            size = (size, size)
-        self.size = size
+        if scale_factor is not None and not isinstance(scale_factor, (Integral, tuple)):
+            raise ValueError('scale_factor must be of integer type or tuple of integer types')
+        self.size = _pair(size)
```
Comment (apaszke) [inline torch/nn/_functions/thnn/upsampling.py]: It should be `_pair(size) if size is not None else None`

File: torch/nn/_functions/thnn/upsampling.py [hunk 2]
```diff
@@ -64,13 +69,29 @@ def backward(self, grad_output):
...
 
+    def __init__(self, size=None, scale_factor=None):
+        super(UpsamplingBilinear2d, self).__init__(size, scale_factor)
+
+        if self.scale_factor is not None:
+            self.scale_factor = _pair(self.scale_factor)
+            # we have to be a tuple at this point
+            try:
+                assert len(self.scale_factor) == 2
+                for i in self.scale_factor:
```
Comment (apaszke) [inline torch/nn/_functions/thnn/upsampling.py]:
```python
assert all(isinstance(s, Integral) and s >= 1 for s in self.scale_factor)
```

File: torch/nn/_functions/thnn/upsampling.py [hunk 3]
```diff
@@ -64,13 +69,29 @@ def backward(self, grad_output):
...
+
     def forward(self, input):
         assert input.dim() == 4
 
-        if self.scale_factor:
+        if self.scale_factor is not None:
             self.output_size = (
-                input.size(2) * self.scale_factor,
-                input.size(3) * self.scale_factor,
+                input.size(2) * self.scale_factor[0],
```
Comment (apaszke) [inline torch/nn/_functions/thnn/upsampling.py]: You need to write a `__setstate__` function for modules that were serialized - `self.scale_factor` won't be a tuple for them, so you need to patch it:
```python
def __setstate__(self, state):
    self.__dict__.update(state)
    self.scale_factor = _tuple(self.scale_factor)
```

File: torch/nn/modules/upsampling.py [hunk 4]
```diff
@@ -110,5 +115,21 @@ class UpsamplingBilinear2d(_UpsamplingBase):
 
     """
 
+    def __init__(self, size=None, scale_factor=None):
+        super(UpsamplingBilinear2d, self).__init__(size, scale_factor)
```
Comment (apaszke) [inline torch/nn/modules/upsampling.py]: As said in an earlier PR - can you move these checks to a helper function (prefixed with an underscore) in `functional.py` and call it from here?

File: torch/nn/modules/upsampling.py [hunk 5]
```diff
@@ -11,8 +11,8 @@ def __init__(self, size=None, scale_factor=None):
         super(_UpsamplingBase, self).__init__()
         if size is None and scale_factor is None:
             raise ValueError('either size or scale_factor should be defined')
-        if scale_factor is not None and not isinstance(scale_factor, Integral):
-            raise ValueError('scale_factor must be of integer type')
+        if scale_factor is not None and not isinstance(scale_factor, (Integral, tuple)):
+            raise ValueError('scale_factor must be of integer type or tuple of integer types')
         self.size = _pair(size)
```
Comment (apaszke) [inline torch/nn/modules/upsampling.py]: I think it also needs that if `self.size is not None`

File: torch/nn/_functions/thnn/upsampling.py [hunk 6]
```diff
@@ -64,13 +69,29 @@ def backward(self, grad_output):
 
 class UpsamplingBilinear2d(_UpsamplingBase):
 
+    def __init__(self, size=None, scale_factor=None):
+        super(UpsamplingBilinear2d, self).__init__(size, scale_factor)
+
+        if self.scale_factor is not None:
+            self.scale_factor = _pair(self.scale_factor)
+            # we have to be a tuple at this point
```
Comment (apaszke) [inline torch/nn/_functions/thnn/upsampling.py]: That comment seems is unnecessary

**Comment by andrewgiessel** [comment]
Thanks for the comments @apaszke! I think I made all the changes you asked for.  Please, let me know if I didn't understand something (in particular I just c/p your `__setstate__()` function).

**Comment by andrewgiessel** [comment]
Please see the comment I made [here](https://github.com/pytorch/pytorch/pull/1348#issuecomment-297800836) on #1348 explaining the most recent commit.

**Review by soumith — APPROVED** [review]

**Comment by soumith** [comment]
thanks Andrew!

**Pull request merged**